diff --git a/distrho/src/DistrhoPluginVST.cpp b/distrho/src/DistrhoPluginVST.cpp
index c77b7af..d062579 100644
--- a/distrho/src/DistrhoPluginVST.cpp
+++ b/distrho/src/DistrhoPluginVST.cpp
@@ -931,12 +931,15 @@ public:
 
             value = value > midRange ? ranges.max : ranges.min;
         }
-        else if (hints & kParameterIsInteger)
+
+        value = ranges.getUnnormalizedValue(value);
+
+        if (hints & kParameterIsInteger)
         {
             value = std::round(value);
         }
 
-        const float realValue(ranges.getUnnormalizedValue(value));
+        const float realValue(value);
         fPlugin.setParameterValue(index, realValue);
 
 #if DISTRHO_PLUGIN_HAS_UI
