diff --git a/distrho/src/DistrhoPluginLV2export.cpp b/distrho/src/DistrhoPluginLV2export.cpp
index c074109..b6b2875 100644
--- a/distrho/src/DistrhoPluginLV2export.cpp
+++ b/distrho/src/DistrhoPluginLV2export.cpp
@@ -422,7 +422,7 @@ void lv2_generate_ttl(const char* const basename)
                     pluginString += "        a lv2:InputPort, lv2:AudioPort ;\n";
 
                 pluginString += "        lv2:index " + String(portIndex) + " ;\n";
-                pluginString += "        lv2:symbol \"lv2_" + port.symbol + "\" ;\n";
+                pluginString += "        lv2:symbol \"" + port.symbol + "\" ;\n";
                 pluginString += "        lv2:name \"" + port.name + "\" ;\n";
 
                 if (port.hints & kAudioPortIsSidechain)
@@ -452,7 +452,7 @@ void lv2_generate_ttl(const char* const basename)
                     pluginString += "        a lv2:OutputPort, lv2:AudioPort ;\n";
 
                 pluginString += "        lv2:index " + String(portIndex) + " ;\n";
-                pluginString += "        lv2:symbol \"lv2_" + port.symbol + "\" ;\n";
+                pluginString += "        lv2:symbol \"" + port.symbol + "\" ;\n";
                 pluginString += "        lv2:name \"" + port.name + "\" ;\n";
 
                 if (port.hints & kAudioPortIsSidechain)
@@ -471,7 +471,7 @@ void lv2_generate_ttl(const char* const basename)
             pluginString += "        a lv2:InputPort, atom:AtomPort ;\n";
             pluginString += "        lv2:index " + String(portIndex) + " ;\n";
             pluginString += "        lv2:name \"Events Input\" ;\n";
-            pluginString += "        lv2:symbol \"lv2_events_in\" ;\n";
+            pluginString += "        lv2:symbol \"events_in\" ;\n";
             pluginString += "        rsz:minimumSize " + String(DISTRHO_PLUGIN_MINIMUM_BUFFER_SIZE) + " ;\n";
             pluginString += "        atom:bufferType atom:Sequence ;\n";
 # if (DISTRHO_PLUGIN_WANT_STATE && DISTRHO_PLUGIN_HAS_UI)
@@ -492,7 +492,7 @@ void lv2_generate_ttl(const char* const basename)
             pluginString += "        a lv2:OutputPort, atom:AtomPort ;\n";
             pluginString += "        lv2:index " + String(portIndex) + " ;\n";
             pluginString += "        lv2:name \"Events Output\" ;\n";
-            pluginString += "        lv2:symbol \"lv2_events_out\" ;\n";
+            pluginString += "        lv2:symbol \"events_out\" ;\n";
             pluginString += "        rsz:minimumSize " + String(DISTRHO_PLUGIN_MINIMUM_BUFFER_SIZE) + " ;\n";
             pluginString += "        atom:bufferType atom:Sequence ;\n";
 # if (DISTRHO_PLUGIN_WANT_STATE && DISTRHO_PLUGIN_HAS_UI)
@@ -510,7 +510,7 @@ void lv2_generate_ttl(const char* const basename)
             pluginString += "        a lv2:OutputPort, lv2:ControlPort ;\n";
             pluginString += "        lv2:index " + String(portIndex) + " ;\n";
             pluginString += "        lv2:name \"Latency\" ;\n";
-            pluginString += "        lv2:symbol \"lv2_latency\" ;\n";
+            pluginString += "        lv2:symbol \"latency\" ;\n";
             pluginString += "        lv2:designation lv2:latency ;\n";
             pluginString += "        lv2:portProperty lv2:reportsLatency, lv2:integer, <" LV2_PORT_PROPS__notOnGUI "> ;\n";
             pluginString += "    ] ;\n\n";
@@ -561,7 +561,7 @@ void lv2_generate_ttl(const char* const basename)
                     String symbol(plugin.getParameterSymbol(i));
 
                     if (symbol.isEmpty())
-                        symbol = "lv2_port_" + String(portIndex-1);
+                        symbol = "port_" + String(portIndex-1);
 
                     pluginString += "        lv2:symbol \"" + symbol + "\" ;\n";
 
